README.md
<p>&nbsp;</p>
<p><img src="https://i.imgur.com/2ROuTJe.png" alt="End Remastered Banner" width="1044" height="290" /></p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;">&nbsp;&nbsp;</span></p>
<p style="text-align: center;"><a href="https://discord.com/invite/D9cxayDNSP"><img src="https://img.shields.io/discord/781617369285263400?label=Discord&amp;logo=Discord&amp;logoColor=ffffff&amp;style=for-the-badge" alt="Discord" /></a> &nbsp;&nbsp; &nbsp; <a title="PayPal" href="https://www.paypal.com/donate/?hosted_button_id=G7P6FV26SRGS2"><img src="https://img.shields.io/badge/PayPal-Donations-009ADC?style=for-the-badge&amp;logo=PayPal" alt="PayPal" width="195" height="28" /></a>&nbsp; &nbsp;&nbsp;&nbsp; <a title="Trailer" href="https://www.youtube.com/watch?v=FMwke39lZu8"><img style="height: 28px; width: 180px;" src="https://img.shields.io/badge/YouTube-Trailer-FF2020?style=for-the-badge&amp;logo=YouTube&amp;logoColor=FF2020" alt="Youtube Trailer" /> </a></p>
<p>&nbsp;</p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;"><strong>End Remastered makes your Minecraft experience more challenging by emphasizing the adventurous side of the game. Basically, 12 new Ender Eyes are added along with a giant castle and a challenging dungeon that replaces Vanilla Strongholds. </strong></span></p>
<p>&nbsp;</p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;">Supported Versions: 1.16.3, 1.16.4, 1.16.5, 1.17.1</span><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;"><br /></span></p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;">Fabric: <a href="https://www.curseforge.com/minecraft/mc-mods/endremastered-fabric">https://www.curseforge.com/minecraft/mc-mods/endremastered-fabric</a></span><span style="font-size: 14px;"><span style="color: #a00000;" data-darkreader-inline-color=""><strong><span style="font-family: helvetica, arial, sans-serif;"><br /></span></strong></span></span></p>
<p>&nbsp;</p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;"><span style="font-size: 24px;"><strong>Our Partners:</strong></span></span><span style="font-size: 14px;"><br /></span></p>
<p>&nbsp;</p>
<p><a href="https://billing.apexminecrafthosting.com/aff.php?aff=4481"><img src="https://i.imgur.com/iVozlrW.png" width="1041" height="289" border="0" /></a></p>
<p>&nbsp;</p>
<p><span style="font-size: 14px;">Thanks to Apex Hosting for partnering with End Remastered!<strong><br />Are you looking for a place to play Minecraft with your friends? </strong>Well, look no further because <span style="color: #000080; --darkreader-inline-color: #7faeff;" data-darkreader-inline-color=""><a style="color: #000080;" href="https://billing.apexminecrafthosting.com/aff.php?aff=4481" data-darkreader-inline-color=""><span style="text-decoration: underline;"><span style="color: #ff6600; text-decoration: underline; --darkreader-inline-color: #ff751a;" data-darkreader-inline-color="">Apex Hosting</span></span></a></span> got you covered! With their great and easy to use Control Panel and their multiple servers spread across the entire globe, you can have your own Minecraft server up and running in about 5 minutes. Use this link: <span style="text-decoration: underline; color: #ff6600; --darkreader-inline-color: #ff751a;" data-darkreader-inline-color="">https://billing.apexminecrafthosting.com/aff.php?aff=4481</span> or click on the image below to get started now!</span></p>
<p>&nbsp;</p>
<p><span style="font-size: 14px;">Want to support us directly, you can also donate through PayPal: <span style="color: #000080; --darkreader-inline-color: #47d1d1;" data-darkreader-inline-color=""><a style="color: #000080; --darkreader-inline-color: #47d1d1;" href="https://www.paypal.com/donate/?hosted_button_id=G7P6FV26SRGS2" data-darkreader-inline-color="">Learn More!</a></span></span></p>
<p><span style="font-size: 24px;">&nbsp;</span></p>
<p><span style="font-size: 24px;"><strong>&nbsp;The Mod Content:</strong></span></p>
<p>&nbsp;</p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;">Exploration is the main focus of End Remastered, and to reach the End and beat the Ender Dragon, you will have to find the 12 custom eyes by exploring vanilla structures and fighting vanilla bosses. Once you have all the eyes, you have the choice to follow them until you reach an End Gate or to trade with a Cartographer to obtain a map to the End Castle. </span><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;"><br /></span></p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;">After reaching one of these structures, you will have to find the hidden portal room, and activate the portal by placing the 12 eyes inside the frames.</span><span style="font-size: 14px;"><span style="font-family: helvetica, arial, sans-serif;"><br /></span></span></p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p><span style="font-size: 24px;"><img src="https://i.imgur.com/PuRYpVn.png" alt="The Eyes" width="1042" height="174" /></span><span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;"><br /></span></p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;">The Eyes are a really important part of End Remastered, just like Vanilla Ender Eyes, they allow you to activate the portal to the End. However, unlike Vanilla Ender Eyes, they are not dropped by Endermen, but must be found by exploring, fighting and mining. For more info, click the spoiler tag below.</span></p>
<p>&nbsp;</p>
<div class="spoiler">
<p><span style="font-size: 14px;"><strong>1. Old Eye - Found in Desert Pyramids</strong> </span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">Legends say this eye once belonged to one of the great Sandworms</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><strong>2. Nether Eye - Found in Nether Fortresses</strong></span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">It was lost by a demon in the battle between the Aether and the Nether</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><strong>3. Cold Eye - Found in Igloos</strong></span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">One of the last artifacts left behind by the Iceologers before they all suddenly vanished from the world</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><strong>4. Rogue Eye - Found in Jungle Pyramids</strong></span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">This eye is believed to have given our ancestors the knowledge of Redstone</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><strong>5. Black Eye - Found in Buried Chests</strong></span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">Pirates say it gave sight to the legendary Black Pearl</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><strong>6. Magical Eye - Dropped by Evokers<br /></strong></span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">Without this eye, the Evoker has no power</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><strong>7. Lost Eye - Found in Mineshafts</strong></span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">It is told that it was forged by the first blacksmiths of time</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><strong>8. Corrupted Eye - Found in Pillager Outposts</strong></span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">The eye of a greedy king faded in solitude. Legend says it will bring infinite fortune to its owner</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><strong>9. Wither Eye - Dropped by The Wither</strong></span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">Has witnessed destruction. Maybe it also witnessed your stuff disappear</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><strong>10. Guardian Eye - Get it by Killing Elder Guardian</strong></span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">It has seen many warriors drown into the depths of the sea</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><strong>11. End Crystal Eye - Get it by crafting it using the End Crystal Fragment, a new ore.</strong></span></p>
<p style="padding-left: 30px;"><span style="font-size: 14px;">It is believed to hold a fragment of the Enderdragon Soul</span></p>
<p><span style="font-size: 14px;"><strong>12. Witch Eye </strong>*new* - <strong>Found in the Custom Witch Huts</strong></span></p>
</div>
<p>&nbsp;</p>
<p>&nbsp;<span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;"><strong>&nbsp;</strong></span></p>
<p><span style="font-size: 14px;"><span style="font-family: trebuchet ms, geneva, sans-serif;"><strong><img src="https://i.imgur.com/0yepcQM.png" alt="End Castle" width="1030" height="172" /></strong></span></span></p>
<p>&nbsp;</p>
<p><span style="font-size: 14px;">The End Castle can be found using a Map (However this can be changed in the Config File) that you can obtain by trading with a cartographer. It is a beautiful and massive structure, that focuses on appearance and atmosphere. Click the spoiler tag below to see pictures of this new structure<strong>.<br /></strong></span></p>
<p>&nbsp;</p>
<div class="spoiler">
<p><span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;"><strong><img src="https://media.discordapp.net/attachments/473982781157015557/780172158319591475/2020-11-22_13.20.40.png?width=789&amp;height=493" width="705" height="441" /></strong></span></p>
<p>&nbsp;</p>
<p><span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;"><strong><img src="https://media.discordapp.net/attachments/473982781157015557/780172155442429982/2020-11-22_13.20.24.png?width=789&amp;height=493" width="705" height="441" /></strong></span></p>
<p>&nbsp;</p>
<p><span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;"><strong><img src="https://cdn.discordapp.com/attachments/473982781157015557/780172151775821832/2020-11-22_13.19.34.png" width="705" height="441" /></strong></span></p>
<p>&nbsp;</p>
<p><span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;"><strong><img src="https://cdn.discordapp.com/attachments/473982781157015557/780171827308920841/2020-11-22_13.png" width="705" height="441" /></strong></span></p>
</div>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p><span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;"><strong><img src="https://i.imgur.com/XW3BeJj.png" alt="End Gate" width="1040" height="174" /><br /></strong></span></p>
<p>&nbsp;</p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;">The End Gate is an extensive dungeon that focuses on exploration, traps, and challenges. If you want to find the hidden portal, you will have to use your creativity to solve 3 difficult puzzles. All rooms are generated randomly, meaning that every End Gate is different, and even the puzzles change from one to the other. It can be found using the Custom Eyes (This can also be changed in the Config File) just like vanilla strongholds. Click the spoiler tag below to see pictures of this new structure.</span></p>
<p>&nbsp;</p>
<div class="spoiler">
<p><span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;"><strong><img src="https://i.imgur.com/8Uqo62n.png" alt="End Fortress" width="748" height="421" /><br /></strong></span></p>
<p><span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;"><strong><img class="transparent" src="https://i.imgur.com/u5G9P4w.png" alt="https://i.imgur.com/u5G9P4w.png" width="750" height="422" /></strong></span></p>
<p>&nbsp;</p>
<p><span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;"><strong><img src="https://i.imgur.com/y8Rnmrg.png" alt="https://i.imgur.com/y8Rnmrg.png" width="749" height="422" /></strong></span></p>
</div>
<p><span style="font-family: trebuchet ms, geneva, sans-serif; font-size: 14px;">&nbsp;</span><span style="font-size: 14px; color: #ff00ff; font-family: trebuchet ms, geneva, sans-serif;" data-darkreader-inline-color=""><strong>&nbsp;</strong></span></p>
<p>&nbsp;<img src="https://i.imgur.com/0RbAkjP.png" alt="The Items" width="1042" height="174" /></p>
<p>&nbsp;</p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;">End Remastered also adds some items to the game, such as new gear and new tools (and a new ore). Click the spoiler button to see the crafting recipe for all the items.</span><span style="font-family: helvetica, arial, sans-serif; font-size: 14px;"> <br /></span></p>
<p>&nbsp;</p>
<div class="spoiler">
<p><span style="font-size: 14px;"><strong>The End Crystal Eye:</strong></span></p>
<p><span style="font-size: 14px;"><img class="transparent" src="https://imgur.com/5mZGuje.png" alt="End Crystal eye Craft" width="257" height="118" /></span></p>
<p>&nbsp;</p>
<p><span style="font-size: 14px;"><strong>The Witch Eye:</strong></span></p>
<p><span style="font-size: 14px;"><img class="transparent" src="https://i.imgur.com/RVwFnAw.png" alt="End Crystal eye Craft" width="257" height="118" /></span></p>
<p>&nbsp;</p>
<p><span style="font-size: 14px;"><strong>The End Crystal Ingot:</strong></span></p>
<p><span style="font-size: 14px;"><img class="transparent" src="https://i.imgur.com/YjEJgee.png" alt="End Crystal eye Craft" width="257" height="118" /></span></p>
<p>&nbsp;</p>
<p><span style="font-size: 14px;"><strong>Armor Crafting:</strong></span></p>
<p><span style="font-size: 14px;"><img class="transparent" src="https://i.imgur.com/QgvrkUZ.png" alt="End Crystal eye Craft" width="257" height="118" /></span></p>
</div>
<p><span style="font-size: 14px;">&nbsp;</span></p>
<p><span style="font-family: helvetica, arial, sans-serif; font-size: 18px;">&nbsp;</span><span style="color: #ffffff;" data-darkreader-inline-color=""><br /></span></p>
<p><span style="font-size: 18px;"><em><strong>Support:</strong></em></span></p>
<p><span style="font-size: 14px;">We always try to be aware of issues in End Remastered, if you experience a bug while playing our mod, please report it in the comments or the "Issues" section to help us keep the experience fun and enjoyable for all players.<strong><br /></strong></span></p>
<p><span style="font-size: 14px;">&nbsp;</span><span style="font-size: 14px;"><br /></span></p>
<p><span style="font-size: 14px;"><span style="color: #3366ff;" data-darkreader-inline-color=""><strong>&nbsp;<a style="color: #3366ff;" href="https://discord.gg/D9cxayDNSP" data-darkreader-inline-color="">CLICK HERE</a> </strong></span><strong>to&nbsp;</strong><strong>join our discord server to speak, chat, report problems o</strong><strong>r play with us :)</strong></span></p>


changelog.txt
Build: 1.16.5-36.2.22 - Tue Dec 14 20:03:13 GMT 2021
	jaredlll08:
		(1.16) Fixed KeyMappings only checking if they conflict with themselves. (#8257)

=========
Build: 1.16.5-36.2.21 - Tue Dec 14 19:50:38 GMT 2021
	pupnewfster:
		Backport #8254 to 1.16.5: Add RenderArmEvent to make overriding just the arm rendering not require copying nearly as much vanilla code (#8299)

=========
Build: 1.16.5-36.2.20 - Fri Dec 10 10:00:01 GMT 2021
	oriondevelopment:
		[CVE-2021-44228]: Update Log4J to fix the security issue inside it (#8270)

=========
Build: 1.16.5-36.2.19 - Wed Nov 24 02:14:10 GMT 2021
	daemonumbra:
		Fix potential issue with BackgroundScanHandler for users with slow disc IO (#8211)

=========
Build: 1.16.5-36.2.18 - Mon Nov 22 21:19:17 GMT 2021
	pupnewfster:
		Backport #8207 to 1.16 (#8209)

=========
Build: 1.16.5-36.2.17 - Mon Nov 22 20:46:52 GMT 2021
	xfacthd:
		[1.16.x] Add ForgeSpawnEggItem to lazily handle EntityTypes (#8150)
		
		* Add ForgeSpawnEggItem to lazily handle EntityTypes
		
		* Remove OnlyIn
		
		* Surround EntityType#spawn() call with try/catch

=========
Build: 1.16.5-36.2.16 - Sat Nov 20 15:45:11 GMT 2021
	Unbekannt1998:
		change the AdvancementProvider to make it possabible for mods to hook into it (#7488)
		
		* back Port changes from #8120

=========
Build: 1.16.5-36.2.15 - Sat Nov 20 15:33:16 GMT 2021
	oriondevelopment:
		Implements Lazy ItemStack Capabilities (#7879)

=========
Build: 1.16.5-36.2.14 - Sat Nov 20 14:52:41 GMT 2021
	xfacthd:
		Pass IModelData through IForgeBakedModel#getModelData() before asking the IBakedModel for the particle texture (#8107)

=========
Build: 1.16.5-36.2.13 - Sat Nov 20 14:30:21 GMT 2021
	xfacthd:
		[1.16.x] Fix blocks with flammable material catching fire from lava when they are marked as non-flammable (#8161)

=========
Build: 1.16.5-36.2.12 - Sat Nov 20 14:17:06 GMT 2021
	xfacthd:
		[1.16] Add accessor to an unmodifiable view of the fullPots map in FlowerPotBlock (#8109)
		
		* Add accessor to an unmodifiable view of the fullPots map in FlowerPotBlock
		
		* Fix non-empty pot check
		
		* Remove exception from getter and add test mod

=========
Build: 1.16.5-36.2.11 - Sat Nov 20 13:56:21 GMT 2021
	championash5357:
		Ensure that all render targets in a PostChain use the same format (#8138)
		
		Fixes #6995 

=========
Build: 1.16.5-36.2.10 - Sat Nov 20 13:42:51 GMT 2021
	championash5357:
		Make Broken Blocks Display Associated FluidState (#8129)
		
		Fixes #7253 for 1.16.x

=========
Build: 1.16.5-36.2.9 - Thu Nov 04 07:09:52 GMT 2021
	gigaherz:
		Implement GUI Stacking feature. (#6927)
		
		* Implement GUI Stacking feature.
		You can now use `mc.pushGuiLayer` and `mc.popGuiLayer` to manage the layers.
		mc.displayGuiScreen behaves such that if called with a non-null screen it replaces the entire stack, and if called with null it closes the entire stack.
		
		* Apply review comment fixes from 1.17 counterpart.
		
		* Revert unintended changes.
		
		* Missed these reverts.

=========
Build: 1.16.5-36.2.8 - Sun Oct 03 16:51:54 GMT 2021
	sciwhiz12:
		Add workaround for pairs of `'` being stripped (#8074)
		
		Backport of #8050

	ititus:
		Fix typo in FluidBlockWrapper (#7880)

=========
Build: 1.16.5-36.2.6 - Mon Sep 27 20:24:54 GMT 2021
	brandon4261:
		Backport MC-209819 (#7923)
		
		Backport fix for mc bug https://bugs.mojang.com/browse/MC-209819

=========
Build: 1.16.5-36.2.5 - Sun Sep 19 23:50:14 GMT 2021
	zml:
		[1.16.5] Bump Mixin to 0.8.4 (#8098)

=========
Build: 1.16.5-36.2.4 - Tue Aug 31 16:32:40 GMT 2021
	djbake101:
		[1.16.5] Backport OnDatapackSyncEvent from 1.17.1 (#7985)

=========
Build: 1.16.5-36.2.3 - Tue Aug 31 15:35:00 GMT 2021
	sciwhiz12:
		[1.16.x] Fix all layers of ItemLayerModel being fullbright (#8040)
		
		* Fix all layers of ItemLayerModel being fullbright
		
		Backport of #8039
		
		* Remove duplicate test resources
		
		These resources have corresponding equivalents in the generated tests
		resources folder.

=========
Build: 1.16.5-36.2.2 - Mon Jul 26 21:37:03 GMT 2021
	onelemonyboi:
		Fix rounding related errors with small entity pathfinding. Closes #7730

=========
Build: 1.16.5-36.2.1 - Sat Jul 24 17:39:23 GMT 2021
	supermartijn642:
		Pass the 'includeName' flag to the ItemTossEvent (#7758)

=========
Build: 1.16.5-36.2.0 - Thu Jul 22 01:45:10 GMT 2021
	lexmanos:
		New 1.16.5 RB.

=========
Build: 1.16.5-36.1.66 - Wed Jul 21 21:21:25 GMT 2021
	sizableshrimp:
		ForgeGradle 5.1 and Gradle 7, which allows development on J16. Runtime still targets J8. (#7877)

=========
Build: 1.16.5-36.1.65 - Tue Jul 20 05:01:10 GMT 2021
	lexmanos:
		Work around javac scoping issue.

	lexmanos:
		Fix LAN worlds where registry entries have been removed.

=========
Build: 1.16.5-36.1.63 - Mon Jul 19 19:54:59 GMT 2021
	william:
		Fix FakePlayers having null `connection` and add FakePlayer test mod (#7733)

=========
Build:  - Sat Jul 17 16:23:44 GMT 2021
	ateranimavis:
		Use '{assets_root}' rather than task output path for userdev config (#7873)
		
		Signed-off-by: AterAnimAvis <AterAnimAvis@gmail.com>

=========
Build: 1.16.5-36.1.61 - Sat Jul 17 11:14:53 GMT 2021
	thelnfamous1:
		Patch several mob classes to let them use modded versions of their ranged weapons (#7845)

	baileyholl:
		Adds a damage float in Lightning Entity and patches Entity to use it. (#7849)

	thesilkminer:
		Fix missing userdev config for datagen runs (#7871)
		
		Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

=========
Build: 1.16.5-36.1.58 - Sat Jul 17 02:45:36 GMT 2021
	awsomesauce239:
		Add Fishing bobber projectile impact event (#7842)

	onelemonyboi:
		Removes misleading Nullable annotation from canCreatureSpawn (#7840)

	onelemonyboi:
		Removes misapplied patch in Entity (#7841)

	onelemonyboi:
		Updates MouseHelper to allow for Horizontal Scroll to be recognized as Vertical Scroll (#7834)

	onelemonyboi:
		Corrects the assigning of mob kills, Fixes MC-121048 (#7836)

=========

src/main/java/com/teamremastered/endrem/config/ERConfig.java
package com.teamremastered.endrem.config;

import com.electronwill.nightconfig.core.file.CommentedFileConfig;
import com.electronwill.nightconfig.core.io.WritingMode;
import com.teamremastered.endrem.EndRemastered;
import net.minecraftforge.common.ForgeConfigSpec;
import net.minecraftforge.fml.ModLoadingContext;
import net.minecraftforge.fml.config.ModConfig;
import net.minecraftforge.fml.loading.FMLPaths;

public class ERConfig {
    private static final ForgeConfigSpec.Builder CONFIG = new ForgeConfigSpec.Builder();
    private static ForgeConfigSpec COMMON_CONFIG;

    // ======   Vanilla Modifications   ======
    public static ERConfigGenericEntry<Boolean> STRONGHOLDS_ENABLED = new ERConfigGenericEntry<>(
            "strongholds",
            "Toggle Vanilla Minecraft Strongholds",
            false);

    public static ERConfigGenericEntry<Boolean> USE_ENDER_EYES_ENABLED = new ERConfigGenericEntry<>(
            "ender_eyes",
            "Toggle Ender Eyes Actions (Throwing and Placing in Portal Frames)",
            false);


    // ======   End Remastered Configuration   ======
    public static ERConfigGenericEntry<String> MONSTER_DIFFICULTY = new ERConfigGenericEntry<>(
            "monster_difficulty",
            "Sets the quantity of mobs in End Remastered Structures: \"peaceful\", \"easy\", \"normal\" or \"hard\"",
            "normal");

    public static ERConfigListEntry WHITELISTED_DIMENSIONS = new ERConfigListEntry(
            "whitelisted_dimensions",
            "Comma-separated list of whitelisted dimensions for End Remastered Structures",
            "[minecraft:overworld]");

    public static ERConfigListEntry END_GATE_BLACKLISTED_BIOMES = new ERConfigListEntry(
            "end_gate_blacklisted_biomes",
            "Comma-separated list of blacklisted biomes for the End Gate",
            "[minecraft:ocean, minecraft:deep_ocean]");

    public static ERConfigListEntry END_GATE_WHITELISTED_BIOME_CATEGORIES = new ERConfigListEntry(
            "end_gate_whitelisted_biome_categories",
            "Comma-separated list of whitelisted biome categories for the End Gate",
            "[plains, jungle, taiga, forest, plains, extreme_hills, mesa, savanna, icy, desert, swamp, mushroom, none]");

    public static ERConfigListEntry END_CASTLE_BLACKLISTED_BIOMES = new ERConfigListEntry(
            "end_castle_blacklisted_biomes",
            "Comma-separated list of blacklisted biomes for the End Castle",
            "[minecraft:bamboo_jungle_hills, minecraft:birch_forest_hills, minecraft:dark_forest_hills, minecraft:giant_spruce_taiga_hills, minecraft:giant_tree_taiga_hills, minecraft:gravelly_mountains, minecraft:jungle_edge, minecraft:jungle_hills, minecraft:modified_gravelly_mountains, minecraft:modified_jungle_edge, minecraft:mountain_edge, minecraft:mountains, minecraft:snowy_mountains, minecraft:snowy_taiga_hills, minecraft:snowy_taiga_mountains, minecraft:taiga_hills, minecraft:taiga_mountains, minecraft:tall_birch_hills, minecraft:wooded_hills, minecraft:wooded_mountains, minecraft:savanna_plateau, minecraft:shattered_savanna_plateau]");

    public static ERConfigListEntry END_CASTLE_WHITELISTED_BIOME_CATEGORIES = new ERConfigListEntry(
            "end_castle_whitelisted_biome_categories",
            "Comma-separated list of whitelisted biome categories for the End Castle",
            "[jungle, taiga, forest, none, plains, icy]");

    public static ERConfigListEntry EYE_STRUCTURE_LIST = new ERConfigListEntry(
            "eye_structure_list",
            "Comma-separated list of structures to locate with End Remastered eyes",
            "[endrem:end_gate]");

    public static ERConfigGenericEntry<Integer> EYE_BREAK_CHANCE = new ERConfigGenericEntry<>(
            "eye_break_chance",
            "Percentage chance of eyes breaking when thrown",
            0);

    public static ERConfigListEntry MAP_STRUCTURE_LIST = new ERConfigListEntry(
            "map_structure_list",
            "Comma-separated list of structures to locate with the End Remastered map",
            "[endrem:end_castle]");

    public static ERConfigListEntry MAP_TRADE_VALUES = new ERConfigListEntry(
            "map_trade_values",
            "Values for the End Remastered map trade: [minPrice, maxPrice, expGiven]",
            "30, 40, 12]");

    public static ERConfigListEntry END_CRYSTAL_ARMOR_STATS = new ERConfigListEntry(
            "end_crystal_armor_stats",
            "Stats for End Remastered's End Crystal armor: [durability factor, defense factor, toughness, knockback resistance, makes piglins neutrals, regen duration (in ticks)]",
            "[33, 1, 2, 0.1, true, 20]");

    public static ERConfigListEntry END_CRYSTAL_TOOLS_STATS = new ERConfigListEntry(
            "end_crystal_tools_stats",
            "Stats for End Remastered's End Crystal tools: [durability, speed, damage bonus]",
            "[1325, 10, 3]");

    public static ERConfigGenericEntry<Integer> END_GATE_DISTANCE = new ERConfigGenericEntry<>(
            "end_gate_distance",
            "Average distance in chunks between End Gates (Set to 0 to disable)",
            85
    );

    public static ERConfigGenericEntry<Integer> END_GATE_SPAWN_DISTANCE = new ERConfigGenericEntry<>(
            "end_gate_spawn_distance",
            "Minimum distance in chunks between End Gates and the spawn",
            130
    );

    public static ERConfigGenericEntry<Integer> END_GATE_SIZE = new ERConfigGenericEntry<>(
            "end_gate_size",
            "Number of pieces generated in End Gates",
            20
    );

    public static ERConfigGenericEntry<Integer> END_GATE_HEIGHT = new ERConfigGenericEntry<>(
            "end_gate_height",
            "Y-Coordinate of the starting piece of the End Gate",
            15
    );

    public static ERConfigGenericEntry<Integer> END_CASTLE_DISTANCE = new ERConfigGenericEntry<>(
            "end_castle_distance",
            "Average distance in chunks between End Castles (Set to 0 to disable)",
            100
    );

    public static ERConfigGenericEntry<Integer> END_CASTLE_SPAWN_DISTANCE = new ERConfigGenericEntry<>(
            "end_castle_spawn_distance",
            "Minimum distance in chunks between End Castles and the spawn",
            188
    );

    public static ERConfigGenericEntry<Boolean> END_CASTLE_TERRAFORMING = new ERConfigGenericEntry<>(
            "end_castle_terraforming",
            "Toggle Whether surrounding land will be modified to conform to the bottom of the structure. \n If set to false, the bottom of the structure will be modified to conform to the surrounding land",
            true);

    public static ERConfigGenericEntry<Integer> ANCIENT_WITCH_HUT_DISTANCE = new ERConfigGenericEntry<>(
            "ancient_witch_hut_distance",
            "Average distance in chunks between Ancient Witch Huts (Set to 0 to disable)",
            25
    );

    static {
        init();
        ModLoadingContext.get().registerConfig(ModConfig.Type.COMMON, COMMON_CONFIG, EndRemastered.CONFIG_FILE);
    }

    private static void init() {
        CONFIG.push(EndRemastered.MOD_ID);

        // Add all the custom config values to the config file
        for (ERConfigGenericEntry<?> ccv : ERConfigGenericEntry.erConfigGenericEntries) {
            ccv.setup(CONFIG);
        }

        CONFIG.pop();
        COMMON_CONFIG = CONFIG.build();
    }

    public static void load() {
        final CommentedFileConfig configData = CommentedFileConfig.builder(FMLPaths.CONFIGDIR.get().resolve(EndRemastered.CONFIG_FILE))
                .sync()
                .autosave()
                .writingMode(WritingMode.REPLACE)
                .build();

        configData.load();
        COMMON_CONFIG.setConfig(configData);
    }
}